Improvements to registration chapter

arthurevans opened the pull request:
Includes changes from #1479 (@mercmobily, the GoogleBot will probably ask you to approve this PR because it includes changes you made in another PR ... Don't ask me why. It just does.)

Changes can be previewed here:
https://init-order-dot-polymer-project.appspot.com/1.0/docs/devguide/registering-elements.html#lifecycle-callbacks

mercmobily commented:
Hi,

Looks fantastic. This big has CLA:NO but I don't have a way to actually sign it off...?

Merc.

mercmobily commented:
I got an email saying:

We found a Contributor License Agreement for you (the sender of this pull request) and all commit authors, but as best as we can tell these commits were authored by someone else. If that's the case, please add them to this pull request and have them confirm that they're okay with these commits being contributed to Google. If we're mistaken and you did author these commits, just reply here to confirm.

I guess all I have to say is "I AM TOTALLY OK"? :D

arthurevans commented:
Thanks Merc!

bicknellr commented on 1.0/docs/devguide/registering-elements.md:
```diff
@@ -177,17 +177,75 @@ all imports in the document have finished loading.
+  </tr>
+  <tr>
+    <td><code>detached</code></td>
+    <td>Called after the element is attached to the document.
```
s/attached to/detached from/

bicknellr replied:
It might be useful to reword these so it's more clear that attached and detached aren't one-off callbacks, but happen every time the element is attached / detached. I think the use of 'Called *after*' above in the explanation of ready kind of primed me to read it that way. 'Called *when*', like in attributeChanged, seems to imply that this can occur multiple times, but it also loses the important detail that it occurs after attachment / detachment. Maybe there's another way to state it.

bicknellr commented on 1.0/docs/devguide/registering-elements.md:
```diff
@@ -177,17 +177,75 @@ all imports in the document have finished loading.
+    </td>
+  </tr>
+  <tr>
+    <td><code>attached</code></td>
```
For elements that are written explicitly in the template (/ initial state) of their container, I assume that attached follows ready, but this could be good to make explicit.

bicknellr replied:
Oh, this is mentioned below in 'Ready callback and local DOM initialization'.

bicknellr commented on 1.0/docs/devguide/registering-elements.md:
```diff
@@ -232,56 +308,104 @@ local DOM when the element is constructed.
 
+Note that while the life cycle callbacks listed above will be called in the
+described order for any given element, the  **initialization timing between
+elements may vary** depending on many factors, including whether or not the
```
Can't you at least rely on the parent being initialized before the descendants in its template?

bicknellr replied:
Once again, explained later (in 'Initialization timing for light DOM children').

bicknellr commented on 1.0/docs/devguide/registering-elements.md:
```diff
@@ -232,56 +308,104 @@ local DOM when the element is constructed.
+
+This means that siblings may become `ready` in any order.
+
+For accessing sibling elements when an element initializes, you can call `async`
```
Is there a way to listen for an element's attached callback, rather than deferring until the next tick?

arthurevans replied:
There is not. I would like to improve this section with better practices, but I don't have anything ready to go here. I

bicknellr commented on 1.0/docs/devguide/registering-elements.md:
```diff
@@ -232,56 +308,104 @@ local DOM when the element is constructed.
+
+In terms of local DOM and initialization timing, local DOM children are created,
+their property values are set as specified in the template, and `ready` is
+called on them _before_ their parent's `ready` callback is called.
```
Haha, this is the opposite of what I expected, but makes more sense after thinking about it. I'm pretty sure I ended up relying on this behavior for a bug fix or two but still forgot.

arthurevans commented:
Updated. Thanks for the comments, @bicknellr! Also updated the information about the registration callbacks based on email convo. with @kevinpschaaf.

robdodson commented:
lgtm
